include/seqan/graph_algorithms/depth_first_search.h

```diff
@@ -121,28 +123,72 @@ void depthFirstSearch(TPredecessorMap & predecessor,
     typedef typename Iterator<TGraph, VertexIterator>::Type TVertexIterator;
     typedef typename VertexDescriptor<TGraph>::Type TVertexDescriptor;
     typedef typename Value<TPredecessorMap>::Type TPredVal;
+    typedef typename Iterator<Graph<TSpec>, AdjacencyIterator>::Type TAdjacencyIterator;
+    enum Task {EXPLORE, FINISH};
+    typedef std::pair<TVertexDescriptor, Task> StackItem;
 
-    // Initialization
+    // Initialization - set each vertex as unvisited and with no predecessor.
     resizeVertexMap(predecessor, g);
     resizeVertexMap(disc, g);
     resizeVertexMap(finish, g);
-    TPredVal nilPred = getNil<TVertexDescriptor>();
-
+    TPredVal nil = getNil<TVertexDescriptor>();
     String<bool> tokenMap;
     resizeVertexMap(tokenMap, g);
     TVertexIterator it(g);
     for(;!atEnd(it);goNext(it)) {
         assignProperty(tokenMap, getValue(it), false);
-        assignProperty(predecessor, getValue(it), nilPred);
+        assignProperty(predecessor, getValue(it), nil);
     }
-
     TSize time = 0;
 
+    // We do the DFS non-recursively using a stack. The stack holds two possible tasks:
+    //   EXPLORE, which means we must follow that vertex's edges
+    //   FINISH, which means the vertex just needs a finish time
+    std::vector<StackItem> vStack;
+
+    // The graph may not be connected, so start at every vertex.
     goBegin(it);
-    for(;!atEnd(it);goNext(it)) {
-        TVertexDescriptor u = getValue(it);
-        if (getProperty(tokenMap, u) == false) {
-            _dfsVisit(g, u, tokenMap, predecessor, disc, finish, time);
+    for(; !atEnd(it); goNext(it)) {
+
+        // If the vertex has already been visited, skip it.
+        TVertexDescriptor v = getValue(it);
+        if (getProperty(tokenMap, v)) {
+            continue;
+        }
+
+        vStack.clear();
+        vStack.push_back(StackItem(v, EXPLORE));
+        while (!vStack.empty()) {
+
+            // Get the vertex and task from the top of the stack.
+            StackItem stackItem = vStack.back();
+            vStack.pop_back();
+            TVertexDescriptor v = stackItem.first;
+
+            if (stackItem.second == FINISH) {
+                assignProperty(finish, v, ++time);
+            }
+
```

Comment:
remove lines between closing ```}``` and ```else if``` statement. You can place the comment after the ```else if``` statement on the same line or behind ```{``` on line down.